Support oci repo/chart:version and add --ensure-version check

benoittgt opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:

Closes: https://github.com/helm/helm/issues/11736

- Allow specifying chart and version in OCI push targets, e.g. oci://ghcr.io/org/repo/mychart:1.2.3.
- Same with `--ensure-version` flag as a safety check. If provided and it differs, the command fails.
- Reject digest references for push.

**Special notes for your reviewer**:

1. Aligns with prior work in #12690, but implements via a small centralized helper and a check-only flag.
2. ~~I am not sure we want to add `--version`, as it seems a little misleading to me. One could expect it to push a specific version. Curious to hear what you think. 🤷 Maybe `:version` is good enough.~~

Tested via:

```sh
$ bin/helm create myapp
Creating myapp

$ sed -i '' 's/^version:.*/version: 0.1.0+meta/' myapp/Chart.yaml

$ bin/helm package myapp
Successfully packaged chart and saved it to: /Users/benoit.tigeot/projects/lifen/helm/myapp-0.1.0+meta.tgz

$ bin/helm push myapp-0.1.0+meta.tgz oci://ghcr.io/benoittgt/my-app 
Pushed: ghcr.io/benoittgt/my-app/myapp:0.1.0_meta
Digest: sha256:e232ddea0dd3f93dfc978df1e3a9db586fce0ef292a7f3f502c494218a8670f3
ghcr.io/benoittgt/my-app/myapp:0.1.0_meta contains an underscore.

OCI artifact references (e.g. tags) do not support the plus sign (+). To support
storing semantic versions, Helm adopts the convention of changing plus (+) to
an underscore (_) in chart version tags when pushing to a registry and back to
a plus (+) when pulling from a registry.

$ bin/helm push myapp-0.1.0+meta.tgz oci://ghcr.io/benoittgt/my-app --ensure-version 0.1.0+meta
Pushed: [...redacted...]

$ bin/helm push myapp-0.1.0+meta.tgz oci://ghcr.io/benoittgt/my-app --ensure-version 0.2.0+meta
Error: specified --ensure-version "0.2.0+meta" does not match chart version "0.1.0+meta"

$ bin/helm push myapp-0.1.0+meta.tgz oci://ghcr.io/benoittgt/my-app:0.1.0+meta
Pushed: [...redacted...]

$ bin/helm push myapp-0.1.0+meta.tgz oci://ghcr.io/benoittgt/my-app:0.2.0+meta
Error: tag "0.2.0+meta" does not match provided chart version "0.1.0+meta"
```

**If applicable**:
- [x] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

benoittgt committed:
Add basic registry BuildPushRef to be used in ocipusher

We want an easy way to check for chart version provided

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Simplify normalized version and use the proper function to validate oci

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Add --version cmd flag to push cmd to validate chart version

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Linting

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Try to make doc less focused on version check

It's hard

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Avoid misleading user with ref.Tag with version dash as underscore on er

Before

```
Error: tag "0.2.0_meta" does not match provided chart version "0.1.0+meta"
```

After

```
Error: tag "0.2.0+meta" does not match provided chart version "0.1.0+meta"
```

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Reorganise tests and remove unnecessary config

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

stevehipwell commented:
> I am not sure we want to add --version, as it seems a little misleading to me. One could expect it to push a specific version. Curious to hear what you think. 🤷 Maybe :version is good enough.

I like the inclusion of the `--version` flag, I don't think it's confusing given the push command it targeting a specific artifact and it's consistent with install where the flag can be used as well as an OCI tag. From an automation perspective keeping the version as a separate input makes the whole flow clearer, e.g. `helm push "${artifact}" "${repository}" --version "${version}"`.

benoittgt committed:
rename --version to --ensure-version to better reflect its purpose

Good suggestion from Scott today during dev meeting

https://docs.google.com/document/d/1d-6xJEx0C78csIYSPKJzRPeWaHG_8W1Hjl72OJggwdc/edit?pli=1&tab=t.0
Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

stevehipwell commented:
@benoittgt is there a reason why this is no longer progressing?

stevehipwell commented:
> @benoittgt is there a reason why this is no longer progressing?

@benoittgt I meant specifically to this PR; I know PRs here can take time to progress.

benoittgt commented:
I will rework on this one.

benoittgt committed:
Reject tagged push refs with mismatched chart name

When the user provides an explicit tag (e.g.
`oci://reg/repo/other:1.2.3`), the repo basename must match
`chartName`. Previously `BuildPushRef` silently appended the
chart name, pushing to a different path than intended.

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt commented:
We should be good now.

stevehipwell reviewed (approved):
LGTM

TerryHowe commented:
Might need to rebase to trigger DCO

benoittgt commented:
Thanks @TerryHowe. DCO seems fine now.

benoittgt committed:
Clarify normalizedVersion is for comparison only

Prevent future refactors from switching the returned tag to
the normalized form; the registry client handles '+' -> '_'.

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

TerryHowe commented on pkg/registry/pushref.go:
```diff
@@ -0,0 +1,62 @@
+)
+
+func BuildPushRef(href, chartName, chartVersion string) (string, error) {
+	ref, err := newReference(href)
```
This needs to work for `helm push chart.tgz oci://localhost:5000` case or that might break people.

TerryHowe commented on pkg/registry/pushref.go:
```diff
@@ -0,0 +1,62 @@
+		return "", fmt.Errorf("tag %q does not match provided chart version %q", normalizedTag, chartVersion)
+	}
+
+	// Ensure repository ends with the chart name once (avoid duplication)
```
While you could argue this is an improvement, dedup is going to have to wait for a major release as that would be a breaking change.

TerryHowe commented on pkg/pusher/ocipusher.go:
```diff
@@ -61,6 +59,16 @@ func (pusher *OCIPusher) push(chartRef, href string) error {
 
+	// If an expected version is specified via --ensure-version, enforce it matches Chart.yaml
+	if pusher.opts.expectedVersion != "" && pusher.opts.expectedVersion != meta.Metadata.Version {
+		return fmt.Errorf("specified --ensure-version %q does not match chart version %q", pusher.opts.expectedVersion, meta.Metadata.Version)
```
```suggestion
		return fmt.Errorf("expected chart version %q does not match chart version %q", pusher.opts.expectedVersion, meta.Metadata.Version)
```

TerryHowe commented on pkg/pusher/ocipusher.go:
```diff
@@ -61,6 +59,16 @@ func (pusher *OCIPusher) push(chartRef, href string) error {
+		return fmt.Errorf("specified --ensure-version %q does not match chart version %q", pusher.opts.expectedVersion, meta.Metadata.Version)
+	}
+
+	ref, err := registry.BuildPushRef(href, meta.Metadata.Name, meta.Metadata.Version)
```
On the dedup topic maybe you could trigger if expectedVersion is set, dedup. That would kind of not be a breaking change and we'd get that logic in there.

TerryHowe commented on pkg/registry/pushref.go:
```diff
@@ -0,0 +1,62 @@
+	"path"
+	"strings"
+)
+
```
Doc comment here

TerryHowe commented on pkg/registry/pushref_test.go:
```diff
@@ -0,0 +1,120 @@
+		if tt.wantErr {
+			t.Run(tt.name, func(t *testing.T) {
+				_, err := BuildPushRef(tt.registry, tt.chart, tt.version)
+				if err == nil {
```
Maybe testify if it isn't too much of a hassle.
